ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Codec.hs

```diff
@@ -141,6 +141,19 @@ codecHandshake versionNumberCodec = mkCodecCborLazyBS encodeMsg decodeMsg
                    | (vNumber, vParams) <- vs'
                    ]
 
+      -- Although `MsgProposeVersions'` shall not be sent, for testing purposes
+      -- it is useful to have an encoder for it.
+      encodeMsg (ServerAgency TokConfirm) (MsgProposeVersions' vs) =
+        let vs' = Map.toAscList vs
+        in
+           CBOR.encodeListLen 2
+        <> CBOR.encodeWord 0
+        <> CBOR.encodeMapLen (fromIntegral $ length vs')
+        <> mconcat [    CBOR.encodeTerm (encodeTerm versionNumberCodec vNumber)
+                     <> CBOR.encodeTerm vParams
+                   | (vNumber, vParams) <- vs'
+                   ]
+
```

Comment:
This seems error prone, it is just a duplication of the encoder for `MsgProposeVersions` . Can you combine them?